src/Interactive/Features/Interactive/Core/InteractiveHost.cs

```diff
@@ -31,6 +31,7 @@ internal sealed partial class InteractiveHost : MarshalByRefObject
         // adjustable for testing purposes
         private readonly int _millisecondsTimeout;
         private const int MaxAttemptsToCreateProcess = 2;
+        private readonly SemaphoreSlim _disposeSemaphore = new SemaphoreSlim(initialCount: 1);
```

Comment:
>1 [](start = 91, length = 1)

Shouldn't this start in non-signaled state (initialCount: 0)? 
Actually, since we only need two states we should use ManualResetEventSlim instead of SemaphoreSlim.

Reply:
One more thing: i think this should be moved to the RemoteService instance and used in its Dispose method. The InteractiveHost itself doesn't really need to know about this -- it's all about disposing the RemoteService.

---
In reply to: [162140846](https://github.com/dotnet/roslyn/pull/24058#discussion_r162140846) [](ancestors = 162140846)